scalafmt-core/shared/src/main/scala/org/scalafmt/internal/FormatWriter.scala

```diff
@@ -376,6 +373,155 @@ class FormatWriter(formatOps: FormatOps) {
+      lazy val noAlignTokens = initStyle.align.tokens.isEmpty &&
+        styleMap.tok2style.values.forall(_.align.tokens.isEmpty)
+      if (locations.length != tokens.length || noAlignTokens)
+        Map.empty[TokenHash, Int]
+      else {
+        var columnShift = 0
+        val finalResult = Map.newBuilder[TokenHash, Int]
+
+        // all blocks must be here, to get final flush
+        val blocks = new mutable.HashMap[Tree, AlignBlock]
+        def createBlock(x: Tree) = blocks.getOrElseUpdate(x, new AlignBlock)
+        var prevAlignContainer: Tree = null
+        var prevBlock: AlignBlock =
+          if (initStyle.align.multiline) null else createBlock(null)
+        val lookupBlock: Tree => AlignBlock =
+          if (!initStyle.align.multiline) _ => prevBlock
+          else (x: Tree) => if (x eq null) prevBlock else createBlock(x)
+        val shouldFlush: Tree => Boolean =
+          if (!initStyle.align.multiline) _ => true
+          else (x: Tree) => x eq prevAlignContainer
+
+        var idx = 0
+        while (idx < locations.length) {
+          var alignContainer: Tree = null
+          val columnCandidates = IndexedSeq.newBuilder[FormatLocation]
+          @tailrec
+          def processLine: FormatLocation = {
+            val floc = locations(idx)
+            idx += 1
+            val ok = !floc.state.split.modification.isNewline
+            if (!ok || floc.formatToken.leftHasNewline) columnShift = 0
+            columnShift += floc.shift
+            if (!ok) floc
+            else {
+              getAlignContainer(floc).foreach { container =>
+                if (alignContainer eq null)
+                  alignContainer = container
+                if (alignContainer eq container)
+                  columnCandidates += floc.copy(
+                    alignContainer = container,
+                    alignHashKey = getAlignHashKey(floc),
+                    shift = floc.state.prev.column + columnShift
+                  )
+              }
+              if (idx < locations.length) processLine else floc
+            }
+          }
+          val location = processLine
+          val candidates = columnCandidates.result()
+          val doubleNewline = location.state.split.modification.newlines > 1
+          val block = lookupBlock(alignContainer)
+          if ((block ne null) && block.nonEmpty) {
+            val prev = block.last
+            val matches = columnMatches(prev, candidates, location.formatToken)
+            if (matches > 0) {
+              // truncate candidates if matches are shorter than both lists
+              val truncate =
+                matches < prev.length && matches < candidates.length
+              if (truncate) {
+                block(block.length - 1) = prev.take(matches)
+                block += candidates.take(matches)
+              } else block += candidates
+            }
+            if (doubleNewline || matches == 0 && shouldFlush(alignContainer))
```

Comment:
this line would need another check for non-multiline case (see comments on `createBlock`).